### PR TITLE
Protects ensureNamespace() Against Race with Contour

### PR DESCRIPTION
**Pull request by danehans**
Previously, `ensureNamespace()` was returning an error during the 2nd pass of contour object reconciliation. This is due to a race between the Contour object being updated by ensureFinalizer() and the namespace create operation. This PR updates the control flow of the function to fix the error case for create and delete operations.

__Before:__
```
2020-09-02T12:33:22.826-0700  INFO  controllers.Contour reconciling {"request": "default/contour-sample"}
2020-09-02T12:33:22.838-0700  INFO  controllers.Contour added finalizer to contour  {"finalizer": "contour.operator.projectcontour.io/finalizer", "namespace": "default", "name": "contour-sample"}
2020-09-02T12:33:22.948-0700  INFO  controllers.Contour created namespace {"name": "projectcontour"}
2020-09-02T12:33:22.948-0700  DEBUG controller  Successfully Reconciled {"reconcilerGroup": "operator.projectcontour.io", "reconcilerKind": "Contour", "controller": "contour", "name": "contour-sample", "namespace": "default"}
2020-09-02T12:33:22.948-0700  INFO  controllers.Contour reconciling {"request": "default/contour-sample"}
2020-09-02T12:33:22.974-0700  ERROR controller  Reconciler error  {"reconcilerGroup": "operator.projectcontour.io", "reconcilerKind": "Contour", "controller": "contour", "name": "contour-sample", "namespace": "default", "error": "failed to ensure namespace projectcontour: failed to create namespace projectcontour: namespaces \"projectcontour\" already exists"}
github.com/go-logr/zapr.(*zapLogger).Error
  /Users/daneyonhansen/code/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
  /Users/daneyonhansen/code/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:237
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
  /Users/daneyonhansen/code/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
  /Users/daneyonhansen/code/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
  /Users/daneyonhansen/code/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
  /Users/daneyonhansen/code/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
  /Users/daneyonhansen/code/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
  /Users/daneyonhansen/code/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90
2020-09-02T12:33:23.974-0700  INFO  controllers.Contour reconciling {"request": "default/contour-sample"}
2020-09-02T12:33:23.975-0700  INFO  controllers.Contour namespace exists; skipped adding  {"name": "projectcontour"}
2020-09-02T12:33:23.975-0700  DEBUG controller  Successfully Reconciled {"reconcilerGroup": "operator.projectcontour.io", "reconcilerKind": "Contour", "controller": "contour", "name": "contour-sample", "namespace": "default"}
```

__With this PR__:
```
2020-09-02T12:47:00.753-0700  INFO  controllers.Contour reconciling {"request": "default/contour-sample"}
2020-09-02T12:47:00.763-0700  INFO  controllers.Contour added finalizer to contour  {"finalizer": "contour.operator.projectcontour.io/finalizer", "namespace": "default", "name": "contour-sample"}
2020-09-02T12:47:00.770-0700  INFO  controllers.Contour created namespace {"name": "projectcontour"}
2020-09-02T12:47:00.770-0700  DEBUG controller  Successfully Reconciled {"reconcilerGroup": "operator.projectcontour.io", "reconcilerKind": "Contour", "controller": "contour", "name": "contour-sample", "namespace": "default"}
2020-09-02T12:47:00.770-0700  INFO  controllers.Contour reconciling {"request": "default/contour-sample"}
2020-09-02T12:47:00.770-0700  DEBUG controller  Successfully Reconciled {"reconcilerGroup": "operator.projectcontour.io", "reconcilerKind": "Contour", "controller": "contour", "name": "contour-sample", "namespace": "default"}
```

/cc @jpeach @Miciah

### CODE REVIEW

**Comment by danehans** [comment]
__Before:__
```
2020-09-02T12:33:22.826-0700  INFO  controllers.Contour reconciling {"request": "default/contour-sample"}
2020-09-02T12:33:22.838-0700  INFO  controllers.Contour added finalizer to contour  {"finalizer": "contour.operator.projectcontour.io/finalizer", "namespace": "default", "name": "contour-sample"}
2020-09-02T12:33:22.948-0700  INFO  controllers.Contour created namespace {"name": "projectcontour"}
2020-09-02T12:33:22.948-0700  DEBUG controller  Successfully Reconciled {"reconcilerGroup": "operator.projectcontour.io", "reconcilerKind": "Contour", "controller": "contour", "name": "contour-sample", "namespace": "default"}
2020-09-02T12:33:22.948-0700  INFO  controllers.Contour reconciling {"request": "default/contour-sample"}
2020-09-02T12:33:22.974-0700  ERROR controller  Reconciler error  {"reconcilerGroup": "operator.projectcontour.io", "reconcilerKind": "Contour", "controller": "contour", "name": "contour-sample", "namespace": "default", "error": "failed to ensure namespace projectcontour: failed to create namespace projectcontour: namespaces \"projectcontour\" already exists"}
github.com/go-logr/zapr.(*zapLogger).Error
  /Users/daneyonhansen/code/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
  /Users/daneyonhansen/code/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:237
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
  /Users/daneyonhansen/code/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
  /Users/daneyonhansen/code/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
  /Users/daneyonhansen/code/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
  /Users/daneyonhansen/code/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
  /Users/daneyonhansen/code/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
  /Users/daneyonhansen/code/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90
2020-09-02T12:33:23.974-0700  INFO  controllers.Contour reconciling {"request": "default/contour-sample"}
2020-09-02T12:33:23.975-0700  INFO  controllers.Contour namespace exists; skipped adding  {"name": "projectcontour"}
2020-09-02T12:33:23.975-0700  DEBUG controller  Successfully Reconciled {"reconcilerGroup": "operator.projectcontour.io", "reconcilerKind": "Contour", "controller": "contour", "name": "contour-sample", "namespace": "default"}
```

__With this PR__:
```
2020-09-02T12:47:00.753-0700  INFO  controllers.Contour reconciling {"request": "default/contour-sample"}
2020-09-02T12:47:00.763-0700  INFO  controllers.Contour added finalizer to contour  {"finalizer": "contour.operator.projectcontour.io/finalizer", "namespace": "default", "name": "contour-sample"}
2020-09-02T12:47:00.770-0700  INFO  controllers.Contour created namespace {"name": "projectcontour"}
2020-09-02T12:47:00.770-0700  DEBUG controller  Successfully Reconciled {"reconcilerGroup": "operator.projectcontour.io", "reconcilerKind": "Contour", "controller": "contour", "name": "contour-sample", "namespace": "default"}
2020-09-02T12:47:00.770-0700  INFO  controllers.Contour reconciling {"request": "default/contour-sample"}
2020-09-02T12:47:00.770-0700  DEBUG controller  Successfully Reconciled {"reconcilerGroup": "operator.projectcontour.io", "reconcilerKind": "Contour", "controller": "contour", "name": "contour-sample", "namespace": "default"}
```

File: controller/contour/namespace.go [hunk 1]
```diff
@@ -35,19 +35,16 @@ const (
...
 	ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: name}}
-	err := r.Client.Get(ctx, types.NamespacedName{Name: ns.Name}, ns)
-	switch {
-	case err == nil:
-		r.Log.Info("namespace exists; skipped adding", "name", ns.Name)
-		return nil
-	case errors.IsNotFound(err):
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: ns.Name}, ns); err != nil {
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("failed to get namespace %s: %v", ns.Name, err)
```
Comment (jpeach) [inline controller/contour/namespace.go:40]: Nit: prefer to use `%w` for errors

https://blog.golang.org/go1.13-errors

**Review by jpeach — CHANGES_REQUESTED** [review]
Can you please also add the commentary from the PR into the body of the commit message? This makes life easier for our future selves; we can see the rationale for the change directly without having to follow breadcrumbs back to the PR.

File: controller/contour/namespace.go (hunk 1 above)
Comment (danehans) [reply controller/contour/namespace.go]: @jpeach commit `d70ed28` changes the verb to `%w`.

**Comment by Miciah** [comment]
On further thought...  Were you consistently seeing the failure?  It looks like a race condition, and I don't see how the code change avoids it (although the refactored code does look cleaner).  I would be inclined to ignore "already exists" errors (just let the reconciler rerun, and the get should succeed the next time and result in a no-op), but you could check with `errors.IsAlreadyExists(err)` and ignore such errors.

File: controller/contour/namespace.go [hunk 2]
```diff
@@ -35,19 +35,16 @@ const (
 // ensureNamespace ensures the namespace for the provided name exists.
 func (r *Reconciler) ensureNamespace(ctx context.Context, name string) error {
 	ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: name}}
-	err := r.Client.Get(ctx, types.NamespacedName{Name: ns.Name}, ns)
-	switch {
-	case err == nil:
-		r.Log.Info("namespace exists; skipped adding", "name", ns.Name)
-		return nil
-	case errors.IsNotFound(err):
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: ns.Name}, ns); err != nil {
```
Comment (stevesloka) [inline controller/contour/namespace.go:38]: Is there still a remaining case where there's an error getting the namespace, but it's different than the `IsNotFound(...)` method case?

**Comment by danehans** [comment]
@Miciah thanks for the feedback in https://github.com/projectcontour/contour-operator/pull/19#issuecomment-692350796. You are correct, I was hitting a race condition between the Contour object being updated by `ensureFinalizer()` and the namespace create operation. Commit `0c041d4` updates `ensureNamespace()` and `ensureNamespaceRemoved()` to address this race condition. PTAL and let me know if you have other ideas for addressing the race condition. cc: @stevesloka

File: controller/contour/namespace.go [hunk 3]
```diff
@@ -58,6 +64,12 @@ func (r *Reconciler) ensureNamespaceRemoved(ctx context.Context, name string) er
 	switch {
 	case err == nil:
 		if err := r.Client.Delete(ctx, ns); err != nil {
+			// protect against a race condition between contour being updated
+			// while the namespace is being deleted.
+			if errors.IsGone(err) {
```
Comment (Miciah) [inline controller/contour/namespace.go:69]:
```suggestion
			if errors.IsNotFound(err) {
```
I've always used `IsNotFound` for this kind of check.  Grepping the Kubernetes codebase turns up only a couple uses of `IsGone`; it looks like it is needed only for an esoteric and deprecated API.

**Comment by danehans** [comment]
@Miciah thanks for the review. Commit `9377428` addressees your https://github.com/projectcontour/contour-operator/pull/19#discussion_r488907139.

File: controller/contour/namespace.go (hunk 2 above)
Comment (danehans) [reply controller/contour/namespace.go]: @stevesloka I reorganized the conditions for both ensure namespace functions, PTAL.

File: controller/contour/namespace.go [hunk 4]
```diff
@@ -42,6 +42,12 @@ func (r *Reconciler) ensureNamespace(ctx context.Context, name string) error {
 		return nil
 	case errors.IsNotFound(err):
 		if err := r.Client.Create(context.TODO(), ns); err != nil {
+			// protect against a race condition between contour being updated
```
Comment (skriss) [inline controller/contour/namespace.go:45]: not essential, but I think you could simplify this even further by not bothering to try getting the namespace first - just try creating it, and if you get an AlreadyExists error, you're good.

File: controller/contour/namespace.go [hunk 5]
```diff
@@ -58,6 +64,12 @@ func (r *Reconciler) ensureNamespaceRemoved(ctx context.Context, name string) er
 	switch {
 	case err == nil:
 		if err := r.Client.Delete(ctx, ns); err != nil {
+			// protect against a race condition between contour being updated
```
Comment (skriss) [inline controller/contour/namespace.go:67]: same thing here - I think you could skip the `Get` call, just try deleting the NS, and if you get an IsNotFound error, that's fine.

**Review by skriss — APPROVED** [review]
I had a couple small suggestions for simplification but I don't consider them blocking and this can be merged as-is.

**Review by stevesloka — APPROVED** [review]
/lgtm

We can merge as-is like @skriss mentioned, or clean up the few nits in a new PR.

**Comment by danehans** [comment]
@skriss commit `66a9e56` addresses your review feedback, PTAL.

**Review by skriss — APPROVED** [review]

**Pull request merged**